src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1504,6 +1530,16 @@ FinalizeRouterPlan(PlannedStmt *localPlan, CustomScan *customScan)
 	/* add original range table list for access permission checks */
 	routerPlan->rtable = list_concat(routerPlan->rtable, localPlan->rtable);
 
+#if PG_VERSION_NUM >= PG_VERSION_16
+
+	/*
+	 * We know that extra remoteScanRangeTableEntry has perminfoindex 0
+	 * therefore we can simply use the perminfos we had in localplan
+	 */
+	Assert(remoteScanRangeTableEntry->perminfoindex == 0);
+	routerPlan->permInfos = localPlan->permInfos;
```

Comment:
I'm confused here. Why don't we follow the same logic as `FinalizeNonRouterPlan`? we concat 2 rtables, and I think we should have a single function that does adjustment of the permInfos for the second list. And, we should be able to call that function here?

Reply:
It's true that we concat two rtable lists, but in this case the first list contains just one element (the remoteScanRangeTableEntry), for which we are sure that its perminfoindex hasn't been set - meaning it's 0. So, nothing further needed here.

In FinalizeNonRouterPlan however, we have no idea what the two rtables contain so we just update everything accordingly.

Reply:
I feel like in the future we'll need this kind of function, and it'd be hard for anyone to know these details so well.

So, I don't feel strongly about this, but sounds pretty useful to me, so please re-consider.